src/halmos/bytevec.py

```diff
@@ -610,6 +611,26 @@ def set_word(self, offset: int, value: Word) -> None:
 
         self.set_slice(offset, offset + 32, value)
 
+    def concretize(self, substitution: dict[BitVecRef, BitVecRef]) -> None:
+        """
+        Replace all top-level symbols in the chunks with their corresponding concrete values, if they exist in the given substitution mapping.
+
+        Note: only top-level symbols are currently replaced, not those within nested symbolic terms.
+        This approach is sufficient for current calldata use cases. Performance impacts should be evaluated if all symbol occurrences need to be replaced.
+        """
+        for offset, chunk in self.chunks.items():
+            chunk_data = chunk.data
+            if not isinstance(chunk, SymbolicChunk) or not is_expr_var(chunk_data):
+                continue
+
+            concrete_chunk_data = substitution.get(chunk_data)
+            if concrete_chunk_data is None:  # could be zero
+                continue
+
+            self.chunks[offset] = ConcreteChunk(
```

Comment:
also are you sure we want to mutate in place vs return a concretized (but different) ByteVec object?

Reply:
i actually considered returning a new object, but worried about performance. it's safer to return a new object though, to be used in different use cases. wdyt?

Reply:
fixed to return a new object: https://github.com/a16z/halmos/pull/364/commits/c18fbc586af2cd024425736df2d22eca9cdd2276